### PR TITLE
Input data generation optimization

### PR DESCRIPTION
**Pull request by toji**
Issue: #1757

<hr>

WIP change, still need to apply the same optimization to `generate_data` in `image_copy.spec.ts` and do a bit more cleanup.

Initial results show that this change takes running the full suite of `copyTextureToTexture` tests **from 71s to 20s on my machine.**

### CODE REVIEW

**Comment by toji** [comment]
Updated `image_copy.spec.ts` now as well. That one was a bit trickier, because one of it's uses of the generated data overlapped with another one, and as a result if the same underlying buffer was reused it's data returned by the first call would get corrupted by the second, but I addressed that by wrapping the generator in a class (shared by `copyTextureToTexture` and `image_copy`) and having a second instance for that use.

With this change in place both sets of tests now take **less than a quarter** of their previous times, and the data generation functions have practically disappeared from the profiler results.

Runtimes (different machine than the previous timing)

|                 |  copyTextureToTexture.spec.ts | image_copy.spec.ts |
|------------|---------------------------------|----------------------|
| Without Patch | 49s | 66s |
| With Patch| 12s | 25s |

File: src/webgpu/api/operation/command_buffer/data_generation.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        this.dataBuffer[i + offset] = ((i ** 3 + i + start) % 251) + 1; // Ensure data is always non-zero
+      }
+
+      this.lastOffset = offset;
+      this.lastStart = start;
+      this.lastByteSize = byteSize;
+    }
+  }
+
+  // Returns a new into the generated data that's the correct length. Because this is a view any
```
Comment (kainino0x) [inline src/webgpu/api/operation/command_buffer/data_generation.ts:51]: new -> new view

File: src/webgpu/api/operation/command_buffer/data_generation.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  private lastOffset = 0;
+  private lastStart = 0;
+  private lastByteSize = 0;
+
+  // Find the nearest power of two greater than or equal to the input value.
+  private nextPowerOfTwo(value: number) {
+    return 1 << (32 - Math.clz32(value - 1));
+  }
+
+  private generateData(byteSize: number, start: number = 0, offset: number = 0) {
```
Comment (kainino0x) [inline src/webgpu/api/operation/command_buffer/data_generation.ts:15]: `start` and `offset` need documentation, they're a bit hard to understand

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts [hunk 3]
```diff
@@ -171,10 +176,13 @@ class ImageCopyTest extends GPUTest {
...
   generateData(byteSize: number, start: number = 0, offset: number = 0): Uint8Array {
+    return altDataGenerator.generateView(byteSize, start, offset);
+
     const arr = new Uint8Array(byteSize);
-    for (let i = 0; i < byteSize; ++i) {
+    arr.set(altDataGenerator.generateView(byteSize, start, offset));
+    /*for (let i = 0; i < byteSize; ++i) {
       arr[i + offset] = (i ** 3 + i + start) % 251;
-    }
+    }*/
```
Comment (kainino0x) [inline src/webgpu/api/operation/command_buffer/image_copy.spec.ts:185]: dead code and commented out code here. Also I think this function is no longer used at all?

**Review by kainino0x — APPROVED** [review]

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts (hunk 3 above)
Comment (toji) [reply src/webgpu/api/operation/command_buffer/image_copy.spec.ts]: Done

File: src/webgpu/api/operation/command_buffer/data_generation.ts (hunk 2 above)
Comment (toji) [reply src/webgpu/api/operation/command_buffer/data_generation.ts]: Done

File: src/webgpu/api/operation/command_buffer/data_generation.ts (hunk 1 above)
Comment (toji) [reply src/webgpu/api/operation/command_buffer/data_generation.ts]: Done

**Comment by austinEng** [comment]
Awesome stuff! I'm not going to have a chance to review this soon (lots to catch up on), so I defer to Kai's review.

**Comment by toji** [comment]
CI failure is unrelated to this PR, so I'm going to go ahead and merge.

**Pull request merged**